smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/RequestCompressionTraitValidator.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.validation.validators;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.RequestCompressionTrait;
+import software.amazon.smithy.model.traits.RequiresLengthTrait;
+import software.amazon.smithy.model.traits.StreamingTrait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * Validates the requestCompression trait.
+ *
+ * - Validates that the operation input has no member with both streaming and
+ *   requiresLength traits applied
+ * - Validate encodings has at least one compression algorithm
+ * - Validate encodings are all supported compression algorithms
+ */
+public final class RequestCompressionTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        Set<OperationShape> shapes = model.getOperationShapesWithTrait(RequestCompressionTrait.class);
+        shapes.forEach(shape -> {
+            validateOperationInput(model, shape, events);
+            validateEncodings(model, shape, events);
+        });
```

Comment:
Personal preference, spell out the loop and avoid creating an extra lambda.

```suggestion
        for (OperationShape shape : shapes) {
            validateOperationInput(model, shape, events);
            validateEncodings(model, shape, events);
        }
```